packages/running-extension/src/kernels.tsx

```diff
@@ -49,6 +58,59 @@ export async function addKernelRunningSessionManager(
   // Wait until the relevant services are ready.
   await Promise.all([kernels.ready, kernelspecs.ready, sessions.ready]);
 
+  async function getUnusedKernels() {
+    // Identifies unused kernels
+    const runningKernels = await KernelAPI.listRunning();
+    return runningKernels.filter(
+      kernel => kernel.connections !== undefined && kernel.connections < 1
+    );
+  }
+
+  async function checkShutdownUnusedEnabled() {
+    shutdownUnusedEnabled = (await getUnusedKernels()).length > 0;
+    commands.notifyCommandChanged(SHUTDOWN_UNUSED_BUTTON_CLASS);
+  }
+
+  commands.addCommand(SHUTDOWN_UNUSED_BUTTON_CLASS, {
+    label: shutdownUnusedLabel,
+    execute: async () => {
+      const unusedKernels = await getUnusedKernels();
+
+      if (unusedKernels.length === 0) {
+        void showDialog({
+          title: trans.__('No Unused Kernels'),
+          body: trans.__('There are no unused kernels to shut down.'),
+          buttons: [Dialog.okButton()]
+        });
+        return;
+      }
+
+      const confirmed = await showDialog({
+        title: shutdownUnusedLabel,
+        body: trans.__(
+          `Are you sure you want to shut down the following unused kernels?\n\n${unusedKernels
+            .map(kernel => kernel.name)
+            .join('\n')}`
+        ),
+        buttons: [
+          Dialog.cancelButton(),
+          Dialog.warnButton({ label: shutdownUnusedLabel })
+        ]
+      });
+
+      if (confirmed.button.accept) {
+        for (const kernel of unusedKernels) {
+          await KernelAPI.shutdownKernel(kernel.id);
+        }
+        Promise.all([kernels.refreshRunning(), sessions.refreshRunning()]);
+      }
+    },
+    isEnabled: () => {
+      shutdownUnusedThrottler.invoke();
```

Comment:
Do we want to have it here, or should this be invoked only after user shut downs the kernels, and in response to `kernels.runningChanged`?

I am worried that it is either not awaited as now so it leads to inconsistent state, or that if it were awaited it would lead to delayed rendering.